samples/order/frontend.go

```diff
@@ -12,19 +12,193 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// TODO(jba): use the response subscription to dynamically update a "My Orders" page.
+
 package main
 
 import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
 	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
 	"gocloud.dev/docstore"
+	_ "gocloud.dev/docstore/memdocstore"
 	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/mempubsub"
+	"gocloud.dev/requestlog"
+	"gocloud.dev/server"
 )
 
-// TODO(jba): implement
-
+// A frontend is a web server that takes image-processing orders.
 type frontend struct {
 	requestTopic *pubsub.Topic
 	responseSub  *pubsub.Subscription
 	bucket       *blob.Bucket
 	coll         *docstore.Collection
 }
+
+var (
+	listTemplate      = template.Must(template.ParseFiles("list.htmlt"))
+	orderFormTemplate = template.Must(template.ParseFiles("order-form.htmlt"))
+)
+
+// run starts the server on port and runs it indefinitely.
+func (f *frontend) run(ctx context.Context, port int) error {
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) { http.ServeFile(w, r, "index.html") })
+	http.HandleFunc("/orders/", wrap(f.listOrders))
```

Comment:
I'd rather not have this abstraction in tutorial-y code. As an alternative consider just spelling the error handling out, or have `listOrders` et.al. do `http.Error` directly. If you really want the wrapper, at least rename it to be more expressive like `wrapHttpError`.

Reply:
I think it's a useful technique, and not very complicated, so I kept it but renamed it.